text/0042-partial-fill-rules.md

```diff
@@ -0,0 +1,316 @@
+
+An example partial fill flow goes like this:
+* Alice (the originator) has an output for 1500 MOB. Alice would like to trade up to 1000 MOB for MEOWB at a price of 10 MEOWB/MOB.
+  Alice doesn't want the trade to happen unless the counterparty takes at least 0.1 MOB.
+* Alice creates an SCI using her 1500 MOB input.
+  * She creates a required output to herself for 500 MOB
+  * She creates a fractional output to herself for 10,000 MEOWB
+  * She creates a fractional change output to herself for 1000 MOB
+  * She sets the `min_fill_value` to 0.1 MOB
+* Bob (the counterparty) sees this quote, and wants to trade 200 MEOWB for 20 MOB (+ transaction fee).
+  * Bob adds this quote to the transaction builder, and indicates that he wants to fill 2% of the quote.
+  * Bob adds an input worth at least 200 MEOWB + transaction fee.
+  * Bob adds a change output with any MEOWB change to himself.
+  * Bob assigns a 20 MOB output to himself.
+
+When the transaction settles, Bob has paid net 200 MEOWB + transaction fee, and received net of 20 MOB.
+Alice has paid a 1500 MOB output, and received the 500 MOB required output, and a 980 MOB fractional output,
+as well as 200 MEOWB from Bob.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The reference level explanation splits into two parts:
+
+1. The [Amount shared secret](#Amount-shared-secret)
+2. The [Partial fill rules schema and semantics](#Partial-fill-rules-schema-and-semantics) of the new `InputRules` fields
+
+Neither of these are legal prior to block version 3. The new SCI fields are legal in block version 3, and the alternate derivation for
+`TxOut::MaskedAmount` must be used in block version 3 and on.
+
+## Amount shared secret
+
+In order to verify that partial fills were performed correctly, we must reveal the amounts of all fractional outputs (which correspond to partial fill outputs) in the `TxPrefix` to the enclave. To do this, we reveal the secret which underlies both the amount commitment and the masked value, so that the enclave can essentially do view-key matching on these TxOut's. A new mechanism is specified to do precisely this, called Amount shared secret.
+
+In order to avoid revealing unnecessary information, we modify the derivation of the value mask and the amount blinding factor of the Pedersen commitment from the TxOut shared secret. We introduce a second step, with a second key called the Amount shared secret.
+
+Before:
+```mermaid
+graph TB
+    A[TxOut SharedSecret] -->|hkdf sha512| B[MemoKey]
+    A[TxOut SharedSecret] -->|blake2b| C[Confirmation Number]
+    A[TxOut SharedSecret] -->|blake2b| D[Amount Blinding Factor]
+    A[TxOut SharedSecret] -->|blake2b| E[Value mask]
+```
+
+After:
+```mermaid
+graph TB
+    A[TxOut SharedSecret] -->|hkdf sha512| B[MemoKey]
+    A -->|blake2b| C[Confirmation Number]
+    A -->|sha512| D[Amount Shared Secret *new*]
```

Comment:
Do you mind adding the explanation here for why we transitioned from blake2b to sha512?

Reply:
I think the reasoning was like:

* `HKDF<Blake2b>` is less well-studied than `HKDF<SHA512>` and the recommendation was not to use `HKDF<Blake2b>` anywhere.
* At one point, `TxOutSharedSecret -> AmountSharedSecret` was an HKDF derivation. However, since there was only one thing to expand, and HKDF involves a second invocation of `SHA512`, it seemed simpler to skip the second `SHA512`, and just use `SHA512` once at this step.

Arguably, we should just use `Blake2b` anywhere are using a hash function a single time, and only use `SHA512` when we have to use `HKDF`. So maybe this should just be changed to `Blake2b` here.

Good question, thanks. Do you have an opinion about which way we should go?

Reply:
Following off-line discussion, we want to make this `Blake2b`, at least to be more consistent with everything else. The only time we need to prefer `Sha512` is when using `HKDF`.